extensions/standard-processors/utils/JoltUtils.h

```diff
@@ -0,0 +1,205 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <vector>
+#include <functional>
+#include <map>
+#include <unordered_map>
+#include <memory>
+#include <compare>
+
+#include "logging/Logger.h"
+#include "utils/gsl.h"
+#include "rapidjson/document.h"
+#include "utils/expected.h"
+#include "utils/StringUtils.h"
+
+namespace org::apache::nifi::minifi::utils::jolt {
+
+class Spec {
+ public:
+  using It = std::string_view::const_iterator;
+
+  struct Context {
+   public:
+    const Context* parent{nullptr};
+
+    std::string path() const {
+      std::string res;
+      if (parent) {
+        res = parent->path();
+      }
+      res.append("/").append(matches.at(0));
+      return res;
+    }
+
+    const Context* find(size_t idx) const {
+      if (idx == 0) return this;
+      if (parent) return parent->find(idx - 1);
+      return nullptr;
+    }
+
+    ::gsl::final_action<std::function<void()>> log(std::function<void(std::shared_ptr<core::logging::Logger>)> on_enter, std::function<void(std::shared_ptr<core::logging::Logger>)> on_exit) const {
```

Comment:
Since this is all in the header anyway, this could be done without `std::function`, just using the function objects themselves directly, by making this a template.

Reply:
changed it to template